scanpy/tests/test_aggregated.py

```diff
@@ -0,0 +1,348 @@
+import anndata as ad
+import scanpy as sc
+import numpy as np
+import pandas as pd
+from scipy.sparse import csr_matrix
+import pytest
+
+
+@pytest.mark.parametrize(
+    'data_key',
+    ['layers', 'obsm', 'varm'],
+)
+@pytest.mark.parametrize(
+    'dim',
+    [
+        'obs',
+        'var',
+    ],
+)
+def test_groupby_different_data_locations(data_key, dim):
+    if (data_key == 'varm' and dim == 'obs') or (data_key == 'obsm' and dim == 'var'):
+        pytest.skip("invalid parameter combination")
+    ax_base = ["A", "B"]
+    ax_groupby = [
+        "v0",
+        "v1",
+        "v2",
+        "w0",
+        "w1",
+        "a1",
+        "a2",
+        "a3",
+        "b1",
+        "b2",
+        "c1",
+        "c2",
+        "d0",
+    ]
+
+    df_groupby = pd.DataFrame(index=pd.Index(ax_groupby, name="cell"))
+    df_groupby["key"] = pd.Categorical([c[0] for c in ax_groupby])
+    df_groupby["key_superset"] = pd.Categorical([c[0] for c in ax_groupby]).map(
+        {'v': 'v', 'w': 'v', 'a': 'a', 'b': 'a', 'c': 'a', 'd': 'a'}
+    )
+    df_groupby["key_subset"] = pd.Categorical([c[1] for c in ax_groupby])
+    df_groupby["weight"] = 2.0
+
+    df_base = pd.DataFrame(index=ax_base)
+
+    X = np.array(
+        [
+            [0, -2],
+            [1, 13],
+            [2, 1],  # v
+            [3, 12],
+            [4, 2],  # w
+            [5, 11],
+            [6, 3],
+            [7, 10],  # a
+            [8, 4],
+            [9, 9],  # b
+            [10, 5],
+            [11, 8],  # c
+            [12, 6],  # d
+        ],
+        dtype=np.float32,
+    )
+    data_dense = X
+    if dim == 'obs':
+        data_sparse_mat_dict = {data_key: {'test': csr_matrix(X)}}
+        adata_sparse = ad.AnnData(
+            **{'obs': df_groupby, 'var': df_base, **data_sparse_mat_dict}
```

Comment:
please do this everywhere:

```suggestion
            obs=df_groupby, var=df_base, **data_sparse_mat_dict
```

Reply:
Got it.